### PR TITLE
Remove proc keyword

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This is the implementation of the [RFC 584](https://github.com/rust-lang/rfcs/pull/584).

### CODE REVIEW

**Comment by steveklabnik** [comment]
This looks good to me, but someone else (maybe @nikomatsakis ) should r+ it

File: src/libsyntax/parse/token.rs [hunk 1]
```diff
@@ -561,7 +561,6 @@ declare_special_idents_and_keywords! {
         (39,                         Virtual,    "virtual");
         (40,                         While,      "while");
         (41,                         Continue,   "continue");
-        (42,                         Proc,       "proc");
         (43,                         Box,        "box");
```
Comment (goffrie) [inline src/libsyntax/parse/token.rs]: Should these numbers be collapsed?

Comment (GuillaumeGomez) [inline src/libsyntax/parse/token.rs]: I didn't see this one. Thanks !

Comment (GuillaumeGomez) [inline src/libsyntax/parse/token.rs]: Hum... Actually that's a pretty good question...

**Comment by GuillaumeGomez** [comment]
It's good now !

**Comment by huonw** [comment]
Maybe we could leave `proc` as a reserved keyword?

**Comment by GuillaumeGomez** [comment]
Why would you want to keep `proc` as a reserved keyword ? When needed, it will just need to be re-implemented and all we will be good.

**Comment by steveklabnik** [comment]
Technically, 're-implementing it' might break backwards compatibility, as you could have named something `proc` in the meantime.

That said, I can see it either way.

**Comment by GuillaumeGomez** [comment]
Well, as you prefer.

**Comment by alexcrichton** [comment]
I would prefer to just leave it as is today. It's a somewhat useful word and we may find use for it one day (especially because it's already reserved). Adding a new keyword forces crates using the keyword to possibly opt-in to having the new keyword semantics, so it's not a free operation.

**Comment by GuillaumeGomez** [comment]
Ok, then i re-add it.

**Comment by GuillaumeGomez** [comment]
I re-added it. Waiting for review.

**Comment by bombless** [comment]
https://github.com/GuillaumeGomez/rust/blob/remove-proc/src/libsyntax/parse/token.rs#L525
(Maybe `proc` keyword should be moved from `strict` to `reserved`)
(Ignore me, I'm just random passer-by)

**Comment by GuillaumeGomez** [comment]
@bombless: Well, I'll wait for others' opinion before doing anything but thanks for your comment !

**Comment by alexcrichton** [comment]
Ah yes, could it be moved to the `reserved` section there as well? Also could you leave the test but update the error message that it outputs?

**Comment by GuillaumeGomez** [comment]
Yes of course. I change that.

**Comment by GuillaumeGomez** [comment]
I move `proc` to reserved keywords.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -1052,7 +1052,7 @@ impl<'a> Parser<'a> {
 
         // examine next token to decide to do
         if self.eat_keyword_noexpect(keywords::Proc) {
-            self.parse_proc_type(lifetime_defs)
+            self.fatal("`proc` keyword isn't used for the moment");
```
Comment (alexcrichton) [inline src/libsyntax/parse/parser.rs]: It's ok to remove this specific handling of the `Proc` keyword, an error should show up regardless somewhere in the parser eventually.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -1523,7 +1523,7 @@ impl<'a> Parser<'a> {
             self.expect(&token::CloseDelim(token::Paren));
             TyTypeof(e)
         } else if self.eat_keyword_noexpect(keywords::Proc) {
-            self.parse_proc_type(Vec::new())
+            self.fatal("`proc` keyword isn't used for the moment");
```
Comment (alexcrichton) [inline src/libsyntax/parse/parser.rs]: Same with this, it's ok to just totally remove this branch.

File: src/test/parse-fail/obsolete-proc.rs [hunk 4]
```diff
@@ -1,17 +0,0 @@
...
-//
-// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
-// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
-// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
-// option. This file may not be copied, modified, or distributed
-// except according to those terms.
-
-// Test that we generate obsolete syntax errors around usages of `proc`.
-
-fn foo(p: proc()) { } //~ ERROR obsolete syntax: the `proc` type
```
Comment (alexcrichton) [inline src/test/parse-fail/obsolete-proc.rs]: Can you leave this test here, but just update this error message?

**Comment by GuillaumeGomez** [comment]
@alexcrichton: I applied your comments.

File: src/libsyntax/parse/parser.rs [hunk 5]
```diff
@@ -2286,10 +2282,7 @@ impl<'a> Parser<'a> {
                     return self.parse_lambda_expr(CaptureByValue);
                 }
                 if self.eat_keyword_noexpect(keywords::Proc) {
-                    let span = self.last_span;
-                    let _ = self.parse_proc_decl();
-                    let _ = self.parse_expr();
-                    return self.obsolete_expr(span, ObsoleteSyntax::ProcExpr);
+                    self.fatal("`proc` keyword isn't used for the moment");
```
Comment (alexcrichton) [inline src/libsyntax/parse/parser.rs]: (this special case can also be removed)

**Comment by GuillaumeGomez** [comment]
So now we're back to my first commit haha. @alexcrichton r? ?

File: src/test/parse-fail/obsolete-proc.rs [hunk 6]
```diff
@@ -10,8 +10,8 @@
 
 // Test that we generate obsolete syntax errors around usages of `proc`.
 
-fn foo(p: proc()) { } //~ ERROR obsolete syntax: the `proc` type
+fn foo(p: proc()) { } //~ ERROR: the `proc` type isn't used for the moment
```
Comment (alexcrichton) [inline src/test/parse-fail/obsolete-proc.rs]: Should this error message be updated? I think the parser modifications you made earlier may have changed the error message (you can check with `make check-stage2-pfail`

Comment (GuillaumeGomez) [inline src/test/parse-fail/obsolete-proc.rs]: Well, I think it's a good thing. Like that it's even more precise.

Comment (alexcrichton) [inline src/test/parse-fail/obsolete-proc.rs]: Ah sorry, I meant that I'm not sure that this error message is still generated. Have you run `make check-stage2-pfail` after the recent modifications?

Comment (GuillaumeGomez) [inline src/test/parse-fail/obsolete-proc.rs]: Ah no, I always stop at stage1. And even if this error message isn't generated, a nice comment is always useful. ^^ Or it wasn't what you meant ?

Comment (alexcrichton) [inline src/test/parse-fail/obsolete-proc.rs]: What I'm saying is that this error message cannot happen an this test needs to be updated if it wants to get past the bots. The `ObsoleteSyntax` changes you made above are never actually called (`ProcExpr` is never referenced and `ProcType` is only called in a dead function).

Could you:

* Delete `ProcExpr` and `ProcType`
* Delete `parse_proc_type`
* Update the error messages in this test

Comment (GuillaumeGomez) [inline src/test/parse-fail/obsolete-proc.rs]: I removed `ProcExpr`, `ProcType` and `parse_proc_type`. I'll run `make check-stage2-pfail` tomorrow.

**Comment by GuillaumeGomez** [comment]
Ready for review !

**Pull request merged**